addon/components/layers/kml-layer.js

```diff
@@ -0,0 +1,149 @@
+/**
+  @module ember-flexberry-gis
+*/
+
+import Ember from 'ember';
+import BaseLayer from '../base-layer';
+
+/* globals Terraformer, omnivore */
+
+/**
+  Kml layer component for leaflet map.
+
+  @class KmlLayerComponent
+  @extends BaseLayerComponent
+*/
+export default BaseLayer.extend({
+  /**
+    Specific option names available on the Layer settings tab.
+  */
+  leafletOptions: [
+    'kmlUrl',
+    'kmlString'
+  ],
+
+  /**
+    Sets leaflet layer's visibility.
+    @method _setLayerOpacity
+    @private
+  */
+  _setLayerOpacity() {
+    let leafletLayer = this.get('_leafletObject');
+    let leafletLayerStyle = Ember.get(leafletLayer, 'options.style');
+    if (Ember.isNone(leafletLayerStyle)) {
+      leafletLayerStyle = {};
+      Ember.set(leafletLayer, 'options.style', leafletLayerStyle);
+    }
+
+    let opacity = this.get('opacity');
+    Ember.set(leafletLayerStyle, 'opacity', opacity);
+    Ember.set(leafletLayerStyle, 'fillOpacity', opacity);
+
+    leafletLayer.setStyle(leafletLayerStyle);
+  },
+
+  /**
+    Creates leaflet layer related to layer type.
+
+    @method createLayer
+    @returns <a href="http://leafletjs.com/reference-1.0.1.html#layer">L.Layer</a>|<a href="https://emberjs.com/api/classes/RSVP.Promise.html">Ember.RSVP.Promise</a>
+    Leaflet layer or promise returning such layer.
+  */
+  createLayer(options) {
+    options = Ember.$.extend(true, {}, this.get('options'), options);
+    let customLayer = L.geoJson(null, options);
+
+    let kmlUrl = this.get('options.kmlUrl');
+    let kmlString = this.get('options.kmlString');
+    Ember.assert('Either "kmlUrl" or "kmlString" should be defined!', Ember.isPresent(kmlUrl) || Ember.isPresent(kmlString));
+
+    if (kmlUrl) {
+      return new Ember.RSVP.Promise((resolve, reject) => {
+        let layer = omnivore.kml(kmlUrl, null, customLayer)
+          .on('ready', (e) => {
+            this.set('_leafletObject', layer);
+            resolve(layer);
+          })
+          .on('error', (e) => {
+            reject(e.error || e);
+          });
+      });
+    }
+
+    if (kmlString) {
```

Comment:
useless condition. If kmlUrl is provided, code will not reach this point. Else, assertion will be thrown.

Reply:
No, assertion will be thrown when kmlUrl and kmlString will be empty both.

Reply:
yep. But if kmlUrl is empty, and kmlString is not - this condition is useless. We already know, that kmlUrl in empty, so kmlString is not empty, as there were no assertion.

Reply:
ah. in that case i agree.